src/workerd/api/web-socket.h

```diff
@@ -140,6 +140,55 @@ class ErrorEvent: public Event {
   void visitForGc(jsg::GcVisitor& visitor);
 };
 
+// The forward declaration is necessary so we can make some
+// WebSocket methods accessible to WebSocketPair via friend declaration.
+class WebSocket;
+
+class WebSocketPair: public jsg::Object {
+public:
+  WebSocketPair(jsg::Ref<WebSocket> first, jsg::Ref<WebSocket> second)
+      : sockets { kj::mv(first), kj::mv(second) } {}
+
+  static jsg::Ref<WebSocketPair> constructor();
+
+  jsg::Ref<WebSocket> getFirst() { return sockets[0].addRef(); }
+  jsg::Ref<WebSocket> getSecond() { return sockets[1].addRef(); }
+
+  JSG_RESOURCE_TYPE(WebSocketPair, CompatibilityFlags::Reader flags) {
+    // TODO(soon): These really should be using an indexed property handler rather
+    // than named instance properties but jsg does not yet have support for that.
+    JSG_READONLY_INSTANCE_PROPERTY(0, getFirst);
+    JSG_READONLY_INSTANCE_PROPERTY(1, getSecond);
```

Comment:
Nit (definitely non-blocking): These could use JSG_LAZY_READONLY_INSTANCE_PROPERTY such that the getter for each is only ever executed once. The returned mapped value will be cached.

```suggestion
    JSG_LAZY_READONLY_INSTANCE_PROPERTY(0, getFirst);
    JSG_LAZY_READONLY_INSTANCE_PROPERTY(1, getSecond);
```